pkg/infra/pulumi_aws/iac/sanitization/sanitizer.ts

```diff
@@ -0,0 +1,195 @@
+    result: string
+    violations: Array<string>
+}
+
+export interface SanitizationRule extends ValidationRule {
+    fix?: FixFunc
+}
+
+type FixFunc = (string) => string
+
+export interface ValidationRule {
+    description: string
+
+    validate(string): boolean
+}
+
+export function sanitize(s: string, options: Partial<SanitizationOptions>): SanitizationResult {
+    let result = s
+    let failedRules = new Array<SanitizationRule>()
+    for (let i = 0; i < (options.maxPasses || 5); i++) {
+        let failedRules = options.rules?.filter((r) => !r.validate(result))
+        failedRules?.forEach((f) => console.debug(f))
+        if (options.minLength != null && result.length < options.minLength) {
+            throw new Error(
+                `The sanitized value, "${result}", is shorter than minLength: ${options.minLength}`
+            )
+        }
+        if (options.maxLength != null && result.length > options.maxLength - reservedCharCount) {
+            result = result.substring(0, options.maxLength - reservedCharCount)
+        }
+        if (failedRules?.length === 0) {
+            return { result: result, violations: [] }
+        }
+        failedRules?.forEach((r) => (result = r.fix?.apply(this, [result]) || result))
+        failedRules = options.rules?.filter((r) => !r.validate(result))
+        failedRules?.forEach((f) => console.debug(f))
+        if (failedRules?.length === 0) {
+            return { result, violations: [] }
+        }
+    }
+    return { result, violations: failedRules.map((r) => r.description) }
+}
+
+export function validate(input: string, options: Partial<SanitizationOptions>) {
+    const violations = doValidate(input, options)
+    if (violations.length > 0) {
+        throw new Error(`Invalid input '${input}':\n\t${violations.join('\n\t')}`)
+    }
+}
+
+function doValidate(input: string, options: Partial<SanitizationOptions>): Array<string> {
+    const violations = new Array<string>()
+    if (options.minLength != null && input.length < options.minLength) {
+        violations.push(`Invalid input: "${input}": length < minLength (${options.minLength})`)
+    }
+    if (options.maxLength != null && input.length > options.maxLength - reservedCharCount) {
+        violations.push(
+            `Invalid input: "${input}": length > maxLength (${
+                options.maxLength - reservedCharCount
+            })`
+        )
+    }
+
+    violations.push(
+        ...(options.rules
+            ?.filter((r) => !r.validate(input))
+            .map((r) => `validation rule violated: ${r.description}`) || [])
+    )
+    return violations
+}
+
+function hashComponent(input: string, maxLength: number): string {
+    maxLength = maxLength > 5 ? 5 : maxLength
+    const hash = sha256.sync(input)
+    return hash.substring(0, maxLength <= input.length ? maxLength : input.length)
+}
+
+export function regexpMatch(
+    description: string,
+    pattern: RegExp,
+    fix: FixFunc | undefined = undefined
```

Comment:
small suggestion: I think all of the calls to this are basically:

```
regexpMatch(desc, /^P$/, (n) => n.replace(/P/g, R))
```

... where `P` is some pattern and `R` is a replacement string. For example:

```
regexpMatch('', /^[\w-]+$/, (n) => n.replace(/[^\w-]/g, '-'))
```

If that's true, I think you could simplify this to just take the pattern and a replace char, and generate the `validate` and `FixFunc` from them:

```
regexpMatch('', /[^\w-]/, '-'))
```

Reply:
That's true for a lot of them, but definitely not all.